mailpoet/lib/Automation/Integrations/MailPoet/Actions/SendWelcomeEmailAction.php

```diff
@@ -0,0 +1,119 @@
+    NewslettersRepository $newslettersRepository,
+    ScheduledTasksRepository $scheduledTasksRepository
+  ) {
+    $this->subscribersRepository = $subscribersRepository;
+    $this->welcomeScheduler = $welcomeScheduler;
+    $this->newslettersRepository = $newslettersRepository;
+    $this->scheduledTasksRepository = $scheduledTasksRepository;
+  }
+
+  public function getKey(): string {
+    return 'mailpoet:sendWelcomeEmail'; // TODO: casing for multi word keys? send-welcome-email? send_welcome_email?
+  }
+
+  public function getName(): string {
+    return __('Send Welcome Email', 'mailpoet');
+  }
+
+  public function run(Workflow $workflow, WorkflowRun $workflowRun, Step $step): void {
+    $subscriberSubject = $workflowRun->getSubjects()['mailpoet:subscriber'] ?? null;
+    if (!$subscriberSubject instanceof SubscriberSubject) {
+      throw InvalidStateException::create()->withMessage(__('No mailpoet:subscriber subject provided.', 'mailpoet'));
+    }
+
+    $segmentSubject = $workflowRun->getSubjects()['mailpoet:segment'] ?? null;
+    if (!$segmentSubject instanceof SegmentSubject) {
+      throw InvalidStateException::create()->withMessage(__('No mailpoet:segment subject provided.', 'mailpoet'));
+    }
+
+    $globalSubscriberStatus = $subscriberSubject->getSubscriberStatusField()->getFactory()();
+    if ($globalSubscriberStatus !== SubscriberEntity::STATUS_SUBSCRIBED) {
+      throw InvalidStateException::create()->withMessage(__(sprintf("Cannot send a welcome email to a subscriber with a global subscription status of '%s'.", $globalSubscriberStatus), 'mailpoet'));
+    }
+
+    $segmentName = $segmentSubject->getNameField()->getFactory()();
+    $idField = $subscriberSubject->getSubscriberIdField();
+    $subscriberId = $idField->getFactory()();
+    if ($subscriberId === null) {
+      throw NotFoundException::create()->withMessage(__(sprintf("Subscriber with ID '%s' not found.", $subscriberId), 'mailpoet'));
+    }
+    
+    $subscriber = $this->subscribersRepository->findOneById($subscriberId);
+    // This is for PHPStan, which doesn't understand that retrieving the ID above means the subscriber entity exists
+    if (!$subscriber instanceof SubscriberEntity) {
+      throw NotFoundException::create()->withMessage(__(sprintf("Subscriber with ID '%s' not found.", $subscriberId), 'mailpoet'));
+    }
+
+    // TODO: Are these necessarily "subscribed" status? Maybe there's a better way to check this
+    $subscriberSegments = $subscriber->getSegments();
+    $subscribedSegmentNames = array_map(function(SegmentEntity $segment) {
+      return $segment->getName();
+    }, $subscriberSegments);
+    
+    if (!in_array($segmentName, $subscribedSegmentNames)) {
+      throw InvalidStateException::create()->withMessage(__(sprintf("Subscriber with ID '%s' is not subscribed to segment '%s'.", $subscriberId, $segmentName), 'mailpoet'));
+    }
+    
+    $welcomeEmailId = (int)$step->getArgs()['welcomeEmailId'];
```

Comment:
Is it a valid assumption that this ID would come from the step?

Are the values for `getArgs` typed, or should we assume they're strings that will need to be cast appropriately?

Reply:
> Is it a valid assumption that this ID would come from the step?

I think so, yes. Again, I didn't really "use" it yet, just drafted, so it may need some changes.

> Are the values for getArgs typed, or should we assume they're strings that will need to be cast appropriately?

Having them saved as JSON, we can consider it typed but PHPStan doesn't know so the cast may be needed anyway for now.